File: java/test/src/main/java/test/Ice/binding/RemoteCommunicatorI.java
// Copyright (c) ZeroC, Inc.

package test.Ice.binding;

import com.zeroc.Ice.Communicator;
import com.zeroc.Ice.Current;
import com.zeroc.Ice.ObjectAdapter;
import com.zeroc.Ice.SocketException;

import test.Ice.binding.Test.RemoteCommunicator;
import test.Ice.binding.Test.RemoteObjectAdapterPrx;
import test.TestHelper;

public class RemoteCommunicatorI implements RemoteCommunicator {
    public RemoteCommunicatorI(TestHelper helper) {
        _helper = helper;
    }

    @Override
    public RemoteObjectAdapterPrx createObjectAdapter(
            String name, String endpts, Current current) {
        int retry = 5;
        while (true) {
            try {
                String endpoints = endpts;
                if (endpoints.indexOf("-p") < 0) {
                    endpoints = _helper.getTestEndpoint(_nextPort++, endpoints);
                }
                Communicator com = current.adapter.getCommunicator();
                com.getProperties().setProperty(name + ".ThreadPool.Size", "1");
                ObjectAdapter adapter =
                        com.createObjectAdapterWithEndpoints(name, endpoints);
                return RemoteObjectAdapterPrx.uncheckedCast(
                        current.adapter.addWithUUID(new RemoteObjectAdapterI(adapter)));
            } catch (SocketException ex) {
                if (--retry == 0) {
                    throw ex;
                }
            }
        }
    }

    @Override
    public void deactivateObjectAdapter(
            RemoteObjectAdapterPrx adapter, Current current) {
        adapter.deactivate(); // Collocated call.
    }

    @Override
    public void shutdown(Current current) {
        current.adapter.getCommunicator().shutdown();
    }

    private final TestHelper _helper;
    private int _nextPort = 10;
}

File: swift/test/Ice/binding/RemoteCommunicatorI.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Ice
import TestCommon

class RemoteCommunicatorI: RemoteCommunicator {
    var _nextPort: Int32 = 10
    var _helper: TestHelper

    init(helper: TestHelper) {
        _helper = helper
    }

    func createObjectAdapter(name: String,
                             endpoints endpts: String,
                             current: Ice.Current) throws -> RemoteObjectAdapterPrx? {
        var retry = 5
        while true {
            do {
                let communicator = current.adapter!.getCommunicator()
                var endpoints = endpts
                if !endpoints.contains("-p") {
                    _nextPort += 1
                    endpoints = _helper.getTestEndpoint(properties: communicator.getProperties(), num: _nextPort,
                                                        prot: endpoints)
                }
                try communicator.getProperties().setProperty(key: "\(name).ThreadPool.Size", value: "1")
                let adapter = try communicator.createObjectAdapterWithEndpoints(name: name, endpoints: endpoints)
                return try uncheckedCast(prx: current.adapter!.addWithUUID(RemoteObjectAdapterI(adapter)),
                                         type: RemoteObjectAdapterPrx.self)
            } catch let ex as Ice.SocketException {
                retry -= 1
                if retry == 0 {
                    throw ex
                }
            }
        }
    }

    func deactivateObjectAdapter(adapter: RemoteObjectAdapterPrx?, current: Ice.Current) throws {
        try adapter!.deactivate() // Collocated call.
    }

    func shutdown(current: Ice.Current) throws {
        current.adapter!.getCommunicator().shutdown()
    }
}
